govulncheck/main.go

```diff
@@ -46,62 +46,103 @@ type Output struct {
+		osvEntry := keyMap.(map[string]interface{})
+		osvEntries[osvEntry["id"].(string)] = osvEntry
+		return nil, false
 	}
 
-	keyMap, exists = data["osv"]
-	if exists {
-		keyMapI := keyMap.(map[string]interface{})
-		return keyMapI, !hasException(config, keyMapI["id"].(string))
+	// Findings are only created for OSV entries found in the _current_ module version.
+	// Meaning, if we have a finding with a particular OSV ID, the current module is affected
+	// by it.
+	keyMap, exists = data["finding"]
+	if !exists {
+		return nil, false
+	}
+	findingMap := keyMap.(map[string]interface{})
+	// Since the findings will potentially be duplicated when multiple traces are found
+	// (e.g. when multiple locations within a module are affected), we de-duplicate here.
+	osvID := findingMap["osv"].(string)
+	if _, exists := osvIDs[osvID]; exists {
+		return nil, false
+	}
+
+	// Add the finding to the list of existing OSV IDs to ensure we do not have duplicates.
+	// Save to do here since we also want to de-duplicate OSV IDs with exceptions.
+	osvIDs[osvID] = struct{}{}
+
+	// If we have an exception based on our exception config, ignore the finding.
+	if hasException(config, osvID) {
+		return nil, false
 	}
-	return nil, false
+
+	// Return the associated OSV entry to the finding. The reason we return the OSV
+	// entry over the finding is because the OSV entry has all the relevant metadata for
+	// the vulnerability, whereas the finding points to a specific trace within the code.
+	return osvEntries[osvID], true
 }
 
 func hasException(config *ExceptionConfig, id string) bool {
 	if exception, ok := config.Exceptions[id]; ok {
-		// Implies this vuln is excluded forever
+		// Implies this vuln is excluded forever.
 		if exception.Until == nil || exception.Until.After(time.Now()) {
 			return true
 		}
 	}
 	return false
 }
 
-func main() {
-	exceptionConfig := readExceptionsFile()
-
-	file, err := os.Open(os.Args[1])
+func collectAffectedVulnerabilities(vulnFile string, exceptionConfig *ExceptionConfig) (*Output, error) {
+	file, err := os.Open(vulnFile)
 	if err != nil {
-		printlnAndExitf("error opening vulns file: %v", err)
-		os.Exit(1)
+		return nil, fmt.Errorf("error opening vuln file: %w", err)
 	}
 	defer func() { _ = file.Close() }()
 
 	output := &Output{
 		Data: make([]map[string]interface{}, 0),
 	}
 	decoder := json.NewDecoder(file)
+	uniqueOSVIDs := map[string]struct{}{}
+	osvEntries := map[string]map[string]interface{}{}
 	for {
 		data := make(map[string]interface{})
 		err := decoder.Decode(&data)
 		if err != nil {
 			if err == io.EOF {
 				break
 			}
-			printlnAndExitf("error reading vulns file: %v", err)
-			os.Exit(1)
+			return nil, fmt.Errorf("error reading vuln file: %w", err)
 		}
 
-		findingMap, valid := parseData(exceptionConfig, data)
+		findingMap, valid := parseData(exceptionConfig, data, uniqueOSVIDs, osvEntries)
 		if !valid {
 			continue
 		}
 		output.Data = append(output.Data, findingMap)
```

Comment:
nitpick: To avoid relying on the order of `osv`s and `finding`s - since we already have a full list of `finding`s and `osv`s in `uniqueOSVIDs, osvEntries` -> we could skip populating output in the loop and just generate output in the end from list of `uniqueOSVIDs`.